### PR TITLE
Basic AI first pass.

### PR DESCRIPTION
**Pull request by norrismiv**

### CODE REVIEW

**Comment by norrismiv** [comment]
This adds the ability for monsters to walk and move, along with being hostile and attempting to attack players.

**Comment by norrismiv** [comment]
First pass fixes added. Runs as intended currently. Can be tweaked if needed.

**Comment by norrismiv** [comment]
more optimizations

File: hybrasyl/Game.cs [hunk 1]
```diff
@@ -368,6 +374,16 @@ public static void Main(string[] args)
...
+            _controlThread = new Thread(_monolithControl.Start);
+
+            //foreach (var control in _monolithControls)
+            //{
+            //    Task.Run(() =>
+            //    {
+            //        control.Start();
+            //    });
+            //}
+
```
Comment (baughj) [inline hybrasyl/Game.cs]: I'd really rather not leave commented code in, in general, anywhere, in PRs. Either use it or take it out.

File: hybrasyl/Monolith.cs [hunk 2]
```diff
@@ -133,52 +136,171 @@ private static void SpawnMonster(Monster monster, Map map)
...
+            while (true)
+            {
+                foreach (var map in _maps)
+                {
+                    //var t = Task.Run(() =>
+                    //{
+                        foreach (var mob in map.EntityTree.GetAllObjects().OfType<Monster>())
+                        {
+                            Evaluate(mob, map);
+                        }
```
Comment (baughj) [inline hybrasyl/Monolith.cs]: Realistically, given what `Evaluate` does, this should be handled using a `HybrasylControlMessage` type, e.g. `ControlMessage_EntityEvaluate` or something. That should then have a defined handler in World.cs (   `SetControlMessageHandlers`) which does all of this (so that the logic runs in the message consumption thread).

Running game logic, especially that which modifies state, outside of the consumer thread, is asking for nightmares. We make no guarantee of thread safety across most of the project because the assumption is that only the consumer thread is (directly) modifying game state.

File: hybrasyl/Objects/Creature.cs [hunk 3]
```diff
@@ -386,12 +386,25 @@ public virtual bool Turn(Direction direction)
...
+                {
+                    var mob = obj as Monster;
+                    var x11 = new ServerPacket(0x11);
+                    x11.WriteUInt32(Id);
+                    x11.WriteByte((byte)direction);
+                    foreach (var user in Map.EntityTree.GetObjects(Map.GetViewport(mob.X, mob.Y)).OfType<User>().ToList())
+                    {
+                        user.Enqueue(x11);
+                    }
+                }
```
Comment (baughj) [inline hybrasyl/Objects/Creature.cs]: Didn't we say anytime we ran into these things or added more we were going to make them use `ServerPacketStructures`? *ducks, runs away*

File: hybrasyl/Monolith.cs [hunk 4]
```diff
@@ -133,52 +136,171 @@ private static void SpawnMonster(Monster monster, Map map)
...
+                            break;
+                        case 1:
+                            monster.Turn(Direction.West);
+                            break;
+                        case 2:
+                            monster.Turn(Direction.North);
+                            break;
+                        case 3:
+                            monster.Turn(Direction.South);
+                            break;
```
Comment (baughj) [inline hybrasyl/Monolith.cs]: I believe this could be simpler:

```
if (monster.ShouldWander)
{
    var nextAction = _random.Next(0, 2);
    var nextMove = (Direction)_random.Next(0, 4);

    if (nextAction == 1)
    {
        monster.Walk(nextMove);
    }
    else
    {
        monster.Turn(nextMove);
    }
}
```

File: hybrasyl/Objects/Monster.cs [hunk 5]
```diff
@@ -35,6 +40,14 @@ public class Monster : Creature, ICloneable
 
         private Spawn _spawn;
 
+        private Creatures.Damage _simpleDamage;
+
+        public int ActionDelay = 800;
+
```
Comment (baughj) [inline hybrasyl/Objects/Monster.cs]: It occurs to me that this is a value that should be expressible in XML.

File: hybrasyl/Objects/Monster.cs [hunk 6]
```diff
@@ -35,6 +40,14 @@ public class Monster : Creature, ICloneable
...
+        public DateTime LastAction { get; set; }
+        public bool IsHostile { get; set; }
+        public bool ShouldWander { get; set; }
+
```
Comment (baughj) [inline hybrasyl/Objects/Monster.cs]: These definitely should be, and they should be using XML values.

File: hybrasyl/Objects/Monster.cs [hunk 7]
```diff
@@ -119,9 +231,282 @@ public override void Attack(Castables.Castable castObject, Creature target)
...
+                                            //facing east, attack east
+                                            rect = new Rectangle(this.X, this.Y, 1 + intent.Radius, 1);
+                                        }
+                                        break;
+                                    case Direction.West:
+                                        {
+                                            //facing west, attack west
+                                            rect = new Rectangle(this.X - intent.Radius, this.Y, intent.Radius, 1);
+                                        }
+                                        break;
```
Comment (baughj) [inline hybrasyl/Objects/Monster.cs]: This looks like it is defining, effectively, an AOI / viewport to be used to select a target? I believe all of this should be refactored to use something like `GetAttackViewport_North/South/etc` with a radius parameter. There is an awful lot of repetition here. Many of these cases are effectively identical.

File: hybrasyl/Objects/Monster.cs [hunk 8]
```diff
@@ -146,6 +531,148 @@ public void Sleep()
             //return to idle state
         }
 
+
+        public override bool Walk(Direction direction)
+        {
```
Comment (baughj) [inline hybrasyl/Objects/Monster.cs]: Why all the duplication here? `VisibleObjects` can and should be able to `Walk`, rather than duplicating this from players.

File: hybrasyl/World.cs [hunk 9]
```diff
@@ -1151,6 +1152,11 @@ private void ControlMessage_SpawnMonster(HybrasylControlMessage message)
             map.InsertCreature(monster);
         }
 
+        private void ControlMessage_MonolithControl(HybrasylControlMessage message)
+        {
+            
+        }
+
```
Comment (baughj) [inline hybrasyl/World.cs]: This does nothing at all. Why is it here?

File: hybrasyl/Game.cs (hunk 1 above)
Comment (norrismiv) [reply hybrasyl/Game.cs]: __

File: hybrasyl/Monolith.cs (hunk 2 above)
Comment (norrismiv) [reply hybrasyl/Monolith.cs]: I agree with this, hence the empty method below.

File: hybrasyl/Monolith.cs (hunk 4 above)
Comment (norrismiv) [reply hybrasyl/Monolith.cs]: Could be for sure.

File: hybrasyl/Objects/Creature.cs (hunk 3 above)
Comment (norrismiv) [reply hybrasyl/Objects/Creature.cs]: Yes. However this was direct copied from the user movement method and modified for monsters.

File: hybrasyl/Objects/Monster.cs (hunk 6 above)
Comment (norrismiv) [reply hybrasyl/Objects/Monster.cs]: These are randomly generated at the time of creation. XML will serve no purpose here.

File: hybrasyl/Objects/Monster.cs (hunk 5 above)
Comment (norrismiv) [reply hybrasyl/Objects/Monster.cs]: This was a stopgap to avoid adding in extra to XML. This actually will likely be removed in the future.

File: hybrasyl/Objects/Monster.cs (hunk 7 above)
Comment (norrismiv) [reply hybrasyl/Objects/Monster.cs]: Effectively Identical, but not identical.  This, again, was code copied from the user method and modified for monsters.

File: hybrasyl/Objects/Monster.cs (hunk 8 above)
Comment (norrismiv) [reply hybrasyl/Objects/Monster.cs]: An Item is a VisibleObject. Should an Item be able to walk?

File: hybrasyl/World.cs (hunk 9 above)
Comment (norrismiv) [reply hybrasyl/World.cs]: See above comment on this.

File: hybrasyl/Objects/Monster.cs (hunk 8 above)
Comment (kojasou) [reply hybrasyl/Objects/Monster.cs]: Yes, some of them have feet.

Comment (norrismiv) [reply hybrasyl/Objects/Monster.cs]: This may be viable at a creature level, however, both user and monster would have to use override methods as is due to differences in the method contents.

**Comment by norrismiv** [comment]
Changes added @baughj

File: hybrasyl/Objects/Monster.cs (hunk 6 above)
Comment (baughj) [reply hybrasyl/Objects/Monster.cs]: IsHostile / ShouldWander are randomly generated? Why? Both of those are behavioral settings that you would want to be able to express in XML.

File: hybrasyl/Objects/Monster.cs (hunk 8 above)
Comment (baughj) [reply hybrasyl/Objects/Monster.cs]: I'd rather have override methods than all this duplication. In particular, the viewport calculation can be removed to a shared function in `Creature`.

File: hybrasyl/World.cs [hunk 10]
```diff
@@ -4366,7 +4467,7 @@ public object ScriptMethod(string name, params object[] args)
 
         private void QueueConsumer()
         {
-            while (!MessageQueue.IsCompleted)
+            while (!MessageQueue.IsAddingCompleted)
```
Comment (baughj) [inline hybrasyl/World.cs]: Why is this changed? `IsCompleted` implies `IsAddingCompleted` plus the queue being empty. In general I feel like we want both of those things, especially for certain types of messages (checkpointing, etc).

File: hybrasyl/Objects/Monster.cs (hunk 5 above)
Comment (baughj) [reply hybrasyl/Objects/Monster.cs]: 👍

File: hybrasyl/Objects/Creature.cs (hunk 3 above)
Comment (baughj) [reply hybrasyl/Objects/Creature.cs]: technically correct

File: hybrasyl/Game.cs (hunk 1 above)
Comment (baughj) [reply hybrasyl/Game.cs]: just junk it

File: hybrasyl/Objects/Monster.cs (hunk 7 above)
Comment (baughj) [reply hybrasyl/Objects/Monster.cs]: Regardless, most of the viewport calculations are redundant and should be made less so.

File: hybrasyl/Monolith.cs [hunk 11]
```diff
@@ -133,55 +137,52 @@ private static void SpawnMonster(Monster monster, Map map)
...
+
+        private static void Evaluate(Monster monster, Map map)
+        {
+            if (!(monster.LastAction < DateTime.Now.AddMilliseconds(-monster.ActionDelay))) return;
+
+            var mapTree = map.EntityTree.GetAllObjects();
+            var mapPlayers = mapTree.Any(x => x is User);
+            if (!mapPlayers) return;
+
+            World.MessageQueue.Add(new HybrasylControlMessage(ControlOpcodes.MonolithControl, monster, map));
```
Comment (baughj) [inline hybrasyl/Monolith.cs]: 👍

**Comment by baughj** [comment]
A handful of things and then I think this is good to be merged.

File: hybrasyl/World.cs (hunk 10 above)
Comment (norrismiv) [reply hybrasyl/World.cs]: This was changed during testing. Will revert.

File: hybrasyl/Objects/Monster.cs (hunk 8 above)
Comment (norrismiv) [reply hybrasyl/Objects/Monster.cs]: The viewport calculation could be an override, however, in both cases of player and monster, the code that currently exists will remain. You can dislike this all you'd like, however, it's not 100% identical between the two, and thus, cannot be made generic.

I'm not going to start refactoring out parts of methods into common methods unless it 100% warrants it. In this case, it does not. It's just a waste of time.

File: hybrasyl/Objects/Monster.cs (hunk 7 above)
Comment (norrismiv) [reply hybrasyl/Objects/Monster.cs]: I'm not going to start refactoring out parts of methods into common methods unless it 100% warrants it. In this case, it does not. It's just a waste of time.

File: hybrasyl/Objects/Monster.cs (hunk 6 above)
Comment (norrismiv) [reply hybrasyl/Objects/Monster.cs]: Both are calculated settings derived at spawn time. They should not exist in XML.

File: hybrasyl/Game.cs (hunk 1 above)
Comment (norrismiv) [reply hybrasyl/Game.cs]: I already did. However, I can find plenty of other places where commented code remains.. if you'd like to start getting technical on this.

Comment (baughj) [reply hybrasyl/Game.cs]: Yeah, the point is, those need to be cleaned up too, and we should not add more.
